Signin page

dan-f opened the pull request:
Fixes #1756 

## Of note

The styling is a best-effort pass on my part. Note that:
  - The color palette doesn't match
  - The branding still says "beta"
  - The button and input components aren't the same

That said, the behaviors should be the same. I'll suggest that we split the styling work into a separate story.

## Testing:
- This should meet the acceptance criteria outlined in #1756 
- To test locally, simply open `src/universal/releaseFlags.js` and flip `newSignIn` to `true`. Then restart your server. When you're done, don't forget to flip the flag back ;)

dan-f commented on src/universal/components/SignInPage/Header.js:
```diff
@@ -0,0 +1,36 @@
+import parabolLogo from 'universal/styles/theme/images/brand/parabol-beta-lockup.svg';
+
+
+const headerStyles = {
```
I used basic inline styles. Happy to change them to use emotion or aphrodite, but this was easiest for development.

dan-f commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,211 @@
+    return <LoadingView />;
+  }
+
+  renderLoadingError = () => {
```
If you want to see this state, try hitting `/signin#blah-blah-not-a-real-oidc-response`.

dan-f commented:
Looking into these failures...

dan-f commented:
Update: failures fixed! Looks like `auth0-js` broke the SSR, but I got around this by dynamically loading it.

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+// A naive implementation taking a querystring-formatted string (beginning in '?'),
+// and returning an object of the key/value pairs.  Will not work for null-valued
+// keys, but works for our own `?redirectTo=/path/to/page` contract.
+const parseSearch = (search: string): Object => {
```
this makes me a little nervous. i think we should at least use `decodeURIComponent` since browsers will automatically encode the string & that could make it barf. The _right_ way to do it would be using the browser native url search params: https://davidwalsh.name/query-string-javascript

dan-f replied:
Well isn't that handy!

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+
+  maybeRedirectToApp = () => {
+    if (this.props.hasSession) {
+      const parsedSearch = parseSearch(this.props.location.search);
```
we previously stored this in redux to keep the url clean. if you like this better i'm cool with it, but we should remove `nextUrl` from the authDuck then

dan-f replied:
Oh, well if this behavior is already implemented, I can simply remove it.

dan-f replied:
Actually, yeah, I prefer the stateless querystring param approach. Removing the nextUrl stuff.

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+  };
+
+  maybeRedirectToApp = () => {
+    if (this.props.hasSession) {
```
nit: maybe extract some variables from this.props at the top?

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+    const {hash} = this.props.location;
+    if (hash) {
+      this.setState({loggingIn: true});
+      this.parseAuthResponse(hash)
```
consider async/await pattern

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+    return new Promise((resolve, reject) => {
+      this.props.webAuth.parseHash({hash}, (err, authResult) => {
+        if (err) {
+          return reject(err);
```
man our linter is messed up. 
the value of resolve/reject should not be returned because you can't use that value, anyways.

dan-f replied:
I'm returning to exit early.

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+    this.setState({loggingIn: false, error: null});
+  };
+
+  saveTokens = (response: ParsedAuthResponse): void => {
```
why plural? it should be a single token (encoded & object form), right?

dan-f replied:
Ah yeah, good catch. Was initially plural because the OIDC response includes an access token as well, but we're not using that.

mattkrick commented on src/universal/components/SignInPage/SignInPage.js:
```diff
@@ -0,0 +1,222 @@
+      responseType: 'token'
+    }, (error) => {
+      this.setState({error});
+      throw error; // Inform raven
```
might be more expressive to call raven.captureException.
Then again, personally i say we don't tell raven. all attempts are logged on auth0, and having a whole bunch of "did i use oauth or my password" failures will just pollute the signal/noise ratio on raven

dan-f replied:
Ok, sure!

mattkrick commented on webpack/webpack.config.dev.js:
```diff
@@ -53,6 +53,8 @@ export default {
     new webpack.HotModuleReplacementPlugin(),
     new webpack.NoEmitOnErrorsPlugin(),
     new webpack.DefinePlugin({
+      __AUTH0_CLIENT_ID__: JSON.stringify(process.env.AUTH0_CLIENT_ID),
```
this will bork in prod since the client is determined by the env, not the build.

mattkrick commented on webpack/webpack.config.dll.js:
```diff
@@ -8,6 +8,7 @@ module.exports = {
     vendors: [
       'aphrodite-local-styles',
+      'auth0-js',
       'auth0-lock',
```
remove?

dan-f replied:
No - we're not there yet! Still need signup and password recovery.

mattkrick commented on webpack/webpack.config.prod.js:
```diff
@@ -20,6 +20,7 @@ const clientInclude = [
 
 
 const vendor = [
+  'auth0-js',
```
down below we can also remove the auth0-lock specific stuff

dan-f replied:
I don't understand what you mean by this comment. Could you clarify?

mattkrick commented on src/universal/components/SignInPage/SignIn.js:
```diff
@@ -0,0 +1,56 @@
+
+type Props = {
+  authProviders: Array<AuthProvider>,
+  error?: boolean,
```
consider `hasError`, error = !!parent.error feels weird.

dan-f replied:
Good call!

dan-f commented on .eslintrc.js:
```diff
@@ -99,6 +99,8 @@ module.exports = {
     ]
   },
   "globals": {
+    "__AUTH0_CLIENT_ID__": true,
```
On it.

dan-f commented on package.json:
```diff
@@ -125,6 +125,7 @@
     "assets-webpack-plugin": "^3.5.1",
     "auth0": "^2.7.0",
+    "auth0-js": "^9.2.3",
     "auth0-lock": "^9.2.2",
```
Not until we also implement sign up and password recovery. Remember that this is merging to `master` behind a feature flag!

dan-f commented on src/universal/components/Action/Action.js:
```diff
@@ -33,7 +33,7 @@ const Action = (props) => {
         }
         {__RELEASE_FLAGS__.newSignIn &&
-          <Route exact path="/signin" component={SignIn} />
+          <Route exact path="/signin" component={SignInPage} />
```
Alright, on this.

dan-f commented:
Hey @mattkrick thank you for the review! I've addressed all your feedback, except for one comment which I didn't understand. Please feel free to merge this if you feel good with it, otherwise let me know if you want to see any other changes.

mattkrick commented on flow-typed/webpackDefinePlugins.js:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Makes flow aware of global variables defined by the webpack define plugin.
```
this is a global attached to the window, but it's not fed by webpack. comment may confuse?

dan-f replied:
Oh true! Yes, oops. I'll rename the module to `flow-typed/globals.js`.

mattkrick commented:
1 minor change to a comment but i won't let that hold this up.
Before it leaves feature flag status i'd like to move styles to emotion just because this is gonna be the most popular page on the website so we wanna make sure it's loading as fast as we can make it.

mattkrick commented:
@dan-f feel free to merge!

dan-f commented:
Thanks! Updated to use emotion. Merging once tests pass.
